packages/sdk/android/src/main/java/com/primerioreactnative/components/manager/klarna/PrimerRNHeadlessUniversalCheckoutKlarnaComponent.kt

```diff
@@ -0,0 +1,345 @@
+                          put(JSONObject(Json.encodeToString(validationStatus.error.toPrimerErrorRN())))
+                        }
+                    )
+                  }
+              )
+            }
+          }
+        }
+      }
+
+    private fun JSONObject.putData(collectableData: KlarnaPaymentCollectableData) {
+        put(
+            "data",
+            JSONObject(
+                Json.encodeToString(
+                    when (collectableData) {
+                        is KlarnaPaymentCollectableData.PaymentOptions -> collectableData.toPaymentOptionsRN()
+                        is KlarnaPaymentCollectableData.FinalizePayment -> collectableData.toFinalizePaymentRN()
+                    }
+                )
+            )
+        )
+    }
+
+    @ReactMethod fun addListener(eventName: String?) = Unit
+
+    @ReactMethod fun removeListeners(count: Int?) = Unit
+
+    @ReactMethod
+    fun onSetPaymentOptions(readableMap: ReadableMap, promise: Promise) {
+        val returnIntentUrl = readableMap.getString("returnIntentUrl")
+        val paymentCategory = readableMap.getMap("paymentCategory") as ReadableMap
+        val json = Json { ignoreUnknownKeys = true }
+        val KlarnaPaymentCategoryRN = json.decodeFromString<KlarnaPaymentCategoryRN>(Json.encodeToString(paymentCategory.toHashMap() as Map<String, String>))
+        
+        val activity = getCurrentActivity()
+
+        if (activity == null) {
+            val exception = ErrorTypeRN.NativeBridgeFailed errorTo MISSING_ACTIVITY_ERROR
+            promise.reject(exception.errorId, exception.description)
+        } else if (klarnaComponent == null) {
+            val exception = ErrorTypeRN.NativeBridgeFailed errorTo UNINITIALIZED_ERROR
+            promise.reject(exception.errorId, exception.description)
+        } else {
+            klarnaComponent?.updateCollectedData(KlarnaPaymentCollectableData.PaymentOptions(
+                context = activity,
+                returnIntentUrl = requireNotNull(returnIntentUrl),
+                paymentCategory = KlarnaPaymentCategoryRN.toKlarnaPaymentCategory()
+            ))
+            promise.resolve(null)
+        }
+    }
+
+    @ReactMethod
+    fun onFinalizePayment(promise: Promise) {
+        if (klarnaComponent == null) {
+            val exception = ErrorTypeRN.NativeBridgeFailed errorTo UNINITIALIZED_ERROR
+            promise.reject(exception.errorId, exception.description)
+        } else {
+            klarnaComponent?.updateCollectedData(KlarnaPaymentCollectableData.FinalizePayment)
+            promise.resolve(null)
+        }
+    }
+
+    private fun JSONObject?.toWritableMap(): WritableMap =
+        this?.let { convertJsonToMap(this) } ?: Arguments.createMap()
+
+    private fun JSONArray?.toWritableArray(): WritableArray =
+        this?.let { convertJsonToArray(this) } ?: Arguments.createArray()
+
+    private fun sendEvent(name: String, data: JSONObject?) {
```

Comment:
[nit]: can we use extension functions from `JsonConverter` class in utils package?

Reply:
Updated